components/prize_pool/commons/prize_pool.lua

```diff
@@ -0,0 +1,467 @@
+		end
+	},
+}
+
+local additional = {
+	[additionalDataTypes.GROUPSCORE] = {
+		field = 'wdl',
+		parse = function (placement, input, context)
+			return input
+		end
+	},
+	[additionalDataTypes.LASTVS] = {
+		field = 'lastvs',
+		parse = function (placement, input, context)
+			return placement:_parseOpponentArgs(input, context.date)
+		end
+	},
+	[additionalDataTypes.LASTVSSCORE] = {
+		field = 'lastvsscore',
+		parse = function (placement, input, context)
+			local scores = Table.mapValues(mw.text.split(input, '-'), tonumber)
+			return {score = scores[1], vsscore = scores[2]}
+		end
+	},
+}
+
+function PrizePool:init(args)
+	self.args = self:_parseArgs(args)
+
+	self.pagename = mw.title.getCurrentTitle().text
+	self.date = PrizePool._getTournamentDate()
+	self.opponentType = self.args.type
+	if self.args.opponentLibrary then
+		Opponent = require('Module:'.. self.args.opponentLibrary)
+	end
+
+	self.options = {}
+	self.prizes = {}
+	self.placements = {}
+
+	return self
+end
+
+function PrizePool:_parseArgs(args)
+	local parsedArgs = Table.deepCopy(args)
+	local typeStruct = Json.parseIfString(args.type)
+
+	PrizePool._assertOpponentStructType(typeStruct)
+
+	parsedArgs.type = typeStruct.type
+
+	return parsedArgs
+end
+
+function PrizePool:create()
+	self.options = self:_readConfig(self.args)
+	self.prizes = self:_readPrizes(self.args)
+	self.placements = self:_readPlacements(self.args)
+
+	if self:_hasUsdPrizePool() then
+		self:setConfig('showUSD', true)
+		self:addPrize('USD', 1)
+	end
+
+	return self
+end
+
+function PrizePool:build()
+	-- TODO
+	mw.logObject(self)
+	return ''
+end
+
+function PrizePool:_readConfig(args)
+	for name, configData in pairs(config) do
+		local value = configData.default
+		if configData.read then
+			value = Logic.nilOr(configData.read(args), value)
+		end
+		self:setConfig(name, value)
+	end
+
+	return self.options
+end
+
+--- Parse the input for available prize types overall.
+function PrizePool:_readPrizes(args)
+	for prizeEnum in pairs(prizeTypes) do
+		local prizeDatum = prizeData[prizeTypes[prizeEnum]]
+		assert(prizeDatum, 'Error: Missmatch between prizeData and prizeTypes. Is a prizeType missing in prizeData?')
+		local fieldName = prizeDatum.header
+		if fieldName then
+			args[fieldName .. '1'] = args[fieldName .. '1'] or args[fieldName]
+			for _, prizeValue, index in Table.iter.pairsByPrefix(args, fieldName) do
+				local data = prizeDatum.headerParse(self, prizeValue, args, index)
+				self:addPrize(prizeEnum, index, data)
+			end
+		end
+	end
+
+	return self.prizes
+end
+
+function PrizePool:_readPlacements(args)
+	local currentPlace = 0
+	for placementIndex = 1, math.huge do
+		if not args[placementIndex] then
+			break
+		end
+
+		local placementInput = Json.parseIfString(args[placementIndex])
+		local placement = Placement(placementInput, currentPlace + 1, self.opponentType)
+
+		currentPlace = placement.placeEnd
+		table.insert(self.placements, placement)
+	end
+
+	return self.placements
```

Comment:
```
	local currentPlace = 0
	return Array.mapIndexes(function(placementIndex)
		if not args[placementIndex] then
			return
		end

		local placementInput = Json.parseIfString(args[placementIndex])
		local Placement(placementInput, currentPlace + 1, self.opponentType)

		currentPlace = placement.placeEnd

		return placement
	end)
```

there's Array.mapIndexes() if you're interested in making use of collection tools. slightly different style

(assign to `self.placements` in caller to reduce mutability)

Reply:
Sure sounds good